.drone.yml

```diff
@@ -59,6 +59,8 @@ pipeline:
     image: golang:1.10 # this step is kept as the lowest version of golang that we support
     pull: true
     commands:
+      - go get
+      - go generate .
```

Comment:
@solderjs this is still unresoloved. Could you remove the `go get` & `go generate .`?

Reply:
Why don't you go ahead and try with go1.10 and then with go1.12 so that you can see the problems.

I remember we had a lengthy discussion about it and I'm not convinced that an understanding was reached.

Reply:
The purpose of this specific CI step is to detect a fail if go generate and go get aren't run. This is a simple smoke test, and running the above will mask potential errors. The following CI step will cover running generate.

Reply:
Try it yourself and you will see.

I moved from using sed to go migrate (the purpose of the PR). The build will fail without go generate because the go code won't run and the files that are generated won't exist.

Reply:
running go generate shouldn't be mandatory, and that's what this step is for. To see if a user can just clone the repo and run go build (on the lowest version of go we support). If this step is failing, the PR should change, not the step.

Reply:
The purpose of the PR is so that the project can be built without requiring outside tooling (i.e. "just works", even on Windows).

Hence, you replace non-go tooling with some Go tooling:
https://github.com/go-gitea/gitea/pull/6567/files#diff-7ddfb3e035b42cd70649cc33393fe32cR1

A workaround for that would be to include a dummy version file that just has 0s in it or something.

However, in order to remove the dependency on non-standard tooling requires some sort of compromise. You can't support Go 1.10, Go 1.12, Windows, Linux, Mac, etc and not have a build step and not have vendor support and not have pre-generated files, and not generate files. Something has to give.

Reply:
Actually, looking back on the code I think the version.go thing is solved by having default values already. I think the issue was more getting it to work with go1.10 and go1.12.

Anyway, please try it for yourself. I don't want to keep doing this back and forth and back and forth and back and forth or PRs and making changes to every single person's suggestion and no one actually taking some project ownership and making a call.

I'm fine with your nits if they're legitimate issues but I'm not going to spend potentially several hours to try to find more ways to finagle things just to have people who aren't even trying it out then let it sit there for another 6 months. It's demoralizing.

Reply:
> A workaround for that would be to include a dummy version file that just has 0s in it or something.

A dummy value would be fine. In fact, the "version" value that currently exists in main.go is overwritten most of the time and is treated as a placeholder. 

> no one actually taking project ownership and making a call

As one of the three project owners of Gitea consider this me making a call that Gitea must be able to be built without having to run go generate.

> I'm fine with your nits if they're legitimate issues

As some one who spends hours a day helping new users onboard onto Gitea, and getting them started with development, I have to consider the weight of adding a new requirement, when some already have difficulty with `go build`. As well as the many downstream projects that depend on our builds being able to be built with go build. Some have their CI just run git clone, and go build with specific LDFLAGS, and that's it, not even using go-bindata to package data in the binary (I'm looking at you Debian. Note for debian people looking at this, please don't see this as a criticisim of how you are doing things, I know you have very specific requirements for packaging things.)

> have people who aren't even trying it out then let it sit there for another 6 months.

I completely sympathize with that, and understand fully. Please don't take what I have said as a criticism of your code, or you personally. I can say that I have tested and run your code works and it works as you have explained, and it is great looking code. As a steward of the project I have to ensure that this does not break any implementation of users downstream. I will put some action behind my words and spend some time this weekend (I am in the commonwealth and we celebrate the queen's birthday so I'm including monday as a part of the weekend), and send a PR to your repo, with what I think will work for users downstream.

> It's demoralizing.

I am truly sorry for this. You are spending your free time contributing to the project, and please know I (and everyone else in the project) do appreciate your contributions.

Reply:
Thank you for the thoughtful response.

I didn't take any of what you said as personal or a question of the quality of my code.

> As one of the three project owners of Gitea consider this me making a call that Gitea must be able to be built without having to run go generate.

This is the kind of thing that I need to know.

If I could get a bulleted list of "these 3 people are decision makers, they know what's needed and they know why" generally along with a specific "These conditions must be true and once they're met, this is ready", I'd be as happy as a clam. I can accomplish specific goals and meet the needs of specific people.

The frustrating thing is the "and one more thing" and a month later "and one more thing" and a month later "and one more thing" yet not knowing who's "one more thing" is important and if or when is "the last thing" when I know that my code is ready to merge: value achieved.

I'm also fine when the answer is "no, this isn't a direction we want to go, it doesn't align with goal X". For example: https://github.com/go-gitea/gitea/pull/6565

With this PR, I do have a particular interest in seeing it go through and I'm willing to spend almost any amount of effort to help it along (because I believe in Go and Go's tooling, and that Windows users should be first-class citizens). I just need to know that the criteria _can_ be met.

For now I'm going to table this until go1.10 is off the table. Until then I don't think this can move forward and accomplish my goals and meet your needs.

If you play around with this and discover something different than what I discovered (which I hope to be the case - I get things wrong just as often as anyone) and you decide to commit the last change or two you need to merge it in, I'll be happy to see the email.

Otherwise it sounds like with what you need (which is clear and makes sense), and what I remember as challenges when submitting this, go1.10 has to drop from support before an agreeable set of conditions can be met that satisfy us both.

Thanks for the feedback and the appreciation. 👍

I'm blunt, brash, and sometimes a little rude, but I also appreciate all of your work and regardless of my clearly-expressed frustration in a moment, I don't have any sort of ill-will either. I want the project to grow and be as great as it can be.

Reply:
Sorry for not responding to this earlier, I went to a 3yo's bday party this weekend, and caught some sort of flu, so I'm a bit out of commission right now. I will respond fully to the above and create the PR once I've recovered.

Reply:
ok. after having to take time off for illness and injury I am finally back and able to get on this.

Reply:
To answer your question about who are the project owners are, you can find the most up-to-date list here: https://github.com/go-gitea/gitea/blob/master/CONTRIBUTING.md#owners

In terms of why there are some decisions that don't make sense, yet we adhere to them (ex. non-yoda conditionals), a lot of it comes down to "most of the codebase already is one way, and consistency usually wins", also I have a hesitance to refactor all the code to a specific way, due to it creating conflicts with other PRs in the line. This isn't to say we can't/won't refactor, it usually is something that has to have a discussion first.

> I'm blunt, brash, and sometimes a little rude

Not in the slightest, there are users who demand stuff, call us all sorts of insults, and expect everyone to drop everything at the drop of the hat. I don't get any of that from you, you contribute, are passionate about the project, and while there may be some (understandable) frustration, you are respectful and I feel genuinely want to know how to better work with the project and to understand any limitations that the project hasn't explicitly defined (such as how decisions are made).

> For now I'm going to table this until go1.10 is off the table.

That's a good idea. Hopefully Gitea1.9 will be out soon, and once that is released we can get rid of Golang 1.10 support.